src/main/java/com/uid2/admin/secret/SaltRotation.java

```diff
@@ -94,86 +125,142 @@ private String calculatePreviousSalt(SaltEntry salt, boolean shouldRotate, long
+            var salt = salts[i];
+            var lastUpdated = salt.lastUpdated();
+            var isInTimeWindow = minLastUpdated <= lastUpdated && lastUpdated < maxLastUpdated;
+            var isRotatable = rotatableSaltIndexes.contains(i);
 
-    private void addIndexesToRotate(List<Integer> entryIndexes,
-                                    SaltEntry[] entries,
-                                    long minLastUpdated,
-                                    long maxLastUpdated,
-                                    int maxIndexes) {
-        final List<Integer> candidateIndexes = IntStream.range(0, entries.length)
-                .filter(i -> isBetween(entries[i].lastUpdated(), minLastUpdated, maxLastUpdated))
-                .boxed()
-                .collect(toList());
-        if (candidateIndexes.size() <= maxIndexes) {
-            entryIndexes.addAll(candidateIndexes);
-            return;
+            if (isInTimeWindow && isRotatable) {
+                candidateIndexes.add(i);
+            }
         }
-        Collections.shuffle(candidateIndexes);
-        candidateIndexes.stream().limit(maxIndexes).forEachOrdered(entryIndexes::add);
+        return candidateIndexes;
     }
 
-    private static boolean isBetween(long t, long minInclusive, long maxExclusive) {
-        return minInclusive <= t && t < maxExclusive;
+    private boolean isRotatable(long nextEffective, SaltEntry salt) {
+        if (this.isRefreshFromEnabled) {
+            if (salt.refreshFrom() == null) { // TODO: remove once refreshFrom is no longer optional
+                return true;
+            }
+            return salt.refreshFrom() == nextEffective;
+        }
+
+        return true;
+    }
+
+    private void logSaltAgeCounts(String logEvent, LocalDate targetDate, Instant nextEffective, SaltEntry[] salts) {
+        var formattedDate = DateTimeFormatter.ofPattern("yyyy-MM-dd").format(targetDate);
+
+        var ages = new HashMap<Long, Long>(); // salt age to count
+        for (var salt : salts) {
+            long age = (nextEffective.toEpochMilli() - salt.lastUpdated()) / DAY_IN_MS;
+            ages.put(age, ages.getOrDefault(age, 0L) + 1);
+        }
+
+        for (var entry : ages.entrySet()) {
+            LOGGER.info("{} target-date={} age={} salts={}", logEvent, formattedDate, entry.getKey(), entry.getValue());
+        }
+    }
+
+    private static SaltEntry[] onlySaltsAtIndexes(SaltEntry[] salts, List<Integer> saltIndexes) {
+        SaltEntry[] selected = new SaltEntry[saltIndexes.size()];
+        for (int i = 0; i < saltIndexes.size(); i++) {
+            selected[i] = salts[saltIndexes.get(i)];
+        }
+        return selected;
     }
 
+    private void logSaltCounts(LocalDate targetDate, Instant nextEffective, SaltEntry[] preRotationSalts, SaltEntry[] postRotationSalts, List<Integer> rotatableSaltIndexes, List<Integer> rotatedSaltIndexes) {
+        var rotatableSalts = onlySaltsAtIndexes(preRotationSalts, rotatableSaltIndexes);
```

Comment:
Can we pass in rotatableSalts, rotatedSalts (and saltsForRefresh) directly into this function?

Reply:
RotatableSalts are saltsForRefresh